TRN2 Meshes and Configurations

apoorvtintin opened the pull request:
This PR adds meshes for TRN2/1 for Fuji models and transformer layer configuration favorable to Neuron.

Neuron supports stacked transformer and GroupedQKVLinear instead of FusedGroupedQKVLinear for Grouped Query Attention (GQA)

This is a newer version of the PR https://github.com/apple/axlearn/pull/885. This PR resolved all comments and requested changes mentioned in the linked PR.

apoorvtintin commented:
Added a ModelConfigModifier that overrides the class for a module. Allowing different model configurations based on Model size and platform.

kelvin-zou commented on axlearn/common/trainer_config_modifier.py:
```diff
@@ -146,6 +147,67 @@ def __call__(self, cfg: SpmdTrainer.Config) -> SpmdTrainer.Config:
+                continue
+            # Here we assume x.y.z format.
+            # One example would be model.decoder.transformer.layer.
+            target_modules = module_name.split(".")
```
Can you try to extract a common util function named something like
`def replace_module_recursive(target_modules:str, config_key: str, target_config)` and make it applied to both here and RematSpecModifier

apoorvtintin replied:
I extractred a helper function, let me know if this looks good

kelvin-zou commented on axlearn/common/trainer_config_modifier_test.py:
```diff
@@ -65,6 +67,26 @@ def test_remat_policy_override(self):
+
+        cfg = cfg_modifier(cfg)
+        # The default StackedTransformerLayer should have changed to RepeatedTransformerLayer
+        self.assertRegex(str(cfg.model.decoder), ".*RepeatedTransformerLayer")
```
Can we do exact match here for the config, something like following: 
```suggestion
        self.assertTrue(cfg.model.decoder is RepeatedTransformerLayer.default_config())
```

kelvin-zou commented on axlearn/common/trainer_config_modifier_test.py:
```diff
@@ -65,6 +67,26 @@ def test_remat_policy_override(self):
+        )
+
+        cfg = cfg_modifier(cfg)
+        # The default StackedTransformerLayer should have changed to RepeatedTransformerLayer
```
Add a test to catch mismatch exception as well?

kelvin-zou reviewed (commented):
Thank you for making such change, overall looks good. A few nit comments.

ruomingp commented on axlearn/common/trainer_config_modifier.py:
```diff
@@ -17,7 +18,42 @@
+def find_target_module(
+    module_name: str, cfg: SpmdTrainer.Config
+) -> tuple[ConfigModifier.Config, ConfigModifier.Config, ConfigModifier.Config]:
+    """Recursively search for the target module matching module_name in provided cfg
```
End comments with .

```suggestion
    """Recursively search for the target module matching module_name in provided cfg.
```

ruomingp commented on axlearn/common/trainer_config_modifier.py:
```diff
@@ -17,7 +18,42 @@
+
+def find_target_module(
+    module_name: str, cfg: SpmdTrainer.Config
+) -> tuple[ConfigModifier.Config, ConfigModifier.Config, ConfigModifier.Config]:
```
Do not return tuples. Return a struct or namedtuple:

https://github.com/apple/axlearn/blob/main/docs/ml_api_style.md#avoid-returning-a-tuple-of-values

ruomingp commented on axlearn/common/trainer_config_modifier.py:
```diff
@@ -17,7 +18,42 @@
+from axlearn.common.utils import HybridMeshShape, MeshShape, PartitionSpec
+
+
+def find_target_module(
```
Does this need to be a public method?

ruomingp commented on axlearn/common/trainer_config_modifier.py:
```diff
@@ -146,6 +175,100 @@ def __call__(self, cfg: SpmdTrainer.Config) -> SpmdTrainer.Config:
+        """
+
+        for module_name, model_cfg in self._model_cfg_modifications.items():
+            # No modification if None
+            if not model_cfg:
+                continue
```
Do we need to support this? Users can always omit the entry.

ruomingp commented on axlearn/common/trainer_config_modifier.py:
```diff
@@ -146,6 +175,100 @@ def __call__(self, cfg: SpmdTrainer.Config) -> SpmdTrainer.Config:
+
+            found_module, parent_module, key_in_parent = find_target_module(module_name, cfg)
+
+            # Copy configurations from the config being replaced on a best effort basis
```
Wait, this behavior is not explained in the class comments. So we are not replacing but merging the configs? Maybe we should support a merge function instead?

apoorvtintin replied:
Yeah the goal is to change the config to a similar module. This means most of the configuration can be reused from before. Essentially replacing the module but merging the config. Let me extract out a merge function.

apoorvtintin replied:
Abstracted out a merge function let me know if more changes are needed for this.

apoorvtintin commented:
@ruomingp Thank you for the review, I have addressed all your comments, please let me know if more changes are needed.

ruomingp commented on axlearn/common/trainer_config_modifier.py:
```diff
@@ -146,6 +186,111 @@ def __call__(self, cfg: SpmdTrainer.Config) -> SpmdTrainer.Config:
+        self._model_cfg_modifications = cfg.model_cfg_modifications
+
+    def _merge_configs(self, target_cfg: ConfigBase, found_module: ConfigBase) -> ConfigBase:
+        """Merge configurations from the config being replaced on a best effort basis.
+        Args:
+            target_cfg: configuration that will replace found_module.
+            found_module: existing configuration whose class will be replaced
+                but it's confguration will be merged with target_cfg.
+
+        Returns:
+            The modified config.
+
+        """
```
Please follow Google style guide strictly in terms of linebreaks.

ruomingp commented on axlearn/common/trainer_config_modifier.py:
```diff
@@ -146,6 +186,111 @@ def __call__(self, cfg: SpmdTrainer.Config) -> SpmdTrainer.Config:
+        for module_name, model_cfg in self._model_cfg_modifications.items():
+            found_module = _find_target_module(module_name, cfg)
+
+            model_cfg = self._merge_configs(model_cfg, found_module)
```
`_merge_configs` actually mutates `model_cfg` IIUC. Is it safe to invoke `__call__` multiple times?

apoorvtintin replied:
Agree, it is possible that a modified config is dropped if a parent config is changed. I can sort the modification modules by string so that we always modify parents first. Let me know if you had another solution in mind.

ruomingp commented on axlearn/common/trainer_config_modifier.py:
```diff
@@ -146,6 +186,111 @@ def __call__(self, cfg: SpmdTrainer.Config) -> SpmdTrainer.Config:
+        self._model_cfg_modifications = cfg.model_cfg_modifications
+
+    def _merge_configs(self, target_cfg: ConfigBase, found_module: ConfigBase) -> ConfigBase:
+        """Merge configurations from the config being replaced on a best effort basis.
```
This comment is vague about how merging actually works. Looks like the rule is:

* Use target_cfg.klass
* Use the field value from found_module if the field exists in both configs (and is not "klass")
* Otherwise keep the value from target_cfg

apoorvtintin replied:
Added a more verbose comment

ruomingp commented on axlearn/common/trainer_config_modifier.py:
```diff
@@ -146,6 +186,111 @@ def __call__(self, cfg: SpmdTrainer.Config) -> SpmdTrainer.Config:
+                (e.g. `model.decoder.transformer.layer`) to a Config.
+        """
+
+        model_cfg_modifications: Required[Dict[str, ConfigBase]] = REQUIRED
```
How about:

```suggestion
        model_cfg_modifications: Dict[str, Callable[[ConfigBase], ConfigBase]] = {}
```

?

That is, the values are config transformer functions.

ruomingp commented on axlearn/common/trainer_config_modifier.py:
```diff
@@ -146,6 +186,111 @@ def __call__(self, cfg: SpmdTrainer.Config) -> SpmdTrainer.Config:
+            The modified trainer config.
+        """
+
+        for module_name, model_cfg in self._model_cfg_modifications.items():
+            found_module = _find_target_module(module_name, cfg)
```
In utils.py we have `get_recursively` and `set_recursively` for `Nested[...]`. I wonder if it will be useful to add corresponding methods to ConfigBase. Then we can do something like:

```suggestion
        for cfg_path, cfg_modification in self._model_cfg_modifications.items():
            child_cfg = cfg.get_recursively(cfg_path)
            child_cfg = cfg_modification(child_cfg, path=cfg_path)
            cfg.set_recursively(cfg_path, value=child_cfg)
```

markblee replied:
I wonder if an alternative (which aims to simplify the `ConfigBase` api) is to do something similar to Python's `sorted`; we allow `utils.get_resursively` to take a value fn:
```
# Default behavior is to use key lookup:
utils.get_recursively(..., value_fn=lambda k,v: v[k])

# Custom behavior can be attribute lookup:
utils.get_recursively(..., value_fn=lambda k,v: getattr(v,k))
```
A benefit is that other non-config instances can also leverage `get_recursively`.

ruomingp replied:
Hi @markblee , maybe we can do this in a follow-up PR?

apoorvtintin commented:
Added a more flexible partition_spec modifier that can modify multiple partition_spec attributes in a single module config.

kelvin-zou commented on axlearn/common/trainer_config_modifier.py:
```diff
@@ -146,6 +186,113 @@ def __call__(self, cfg: SpmdTrainer.Config) -> SpmdTrainer.Config:
+                (e.g. `model.decoder.transformer.layer`) to a Config.
+        """
+
+        model_cfg_modifications: Dict[str, Callable[[ConfigBase], ConfigBase]] = {}
```
There is a chain modifier already, should this be something simpler, saying taking two arguments, 
1. target_module
2. modified_config
If you rely on chain, you don't really need a dict here.

apoorvtintin replied:
I felt the function of chain modifier is to chain different types of modifiers together. We have a tradeoff here I can certainly use the same ModelConfigModifier multiple times but that will make the config more verbose, while it will make the modifier simpler. I am open to both approaches.

kelvin-zou commented on axlearn/common/trainer_config_modifier.py:
```diff
@@ -17,7 +18,53 @@
+    key_in_parent = None
+    parent_module = None
+
+    for target_module in target_modules:
```
nit, maybe name `target_module` to `target_module_key` since it is more approriate?

kelvin-zou reviewed (commented):
Mostly lgtm, some minor comments.

apoorvtintin commented:
@ruomingp I addressed your comments, please let me know if anymore changes are needed.

markblee commented on axlearn/experiments/text/gpt/fuji.py:
```diff
@@ -151,6 +155,60 @@ def get_trainer_kwargs(
+    if version != Version.V1:
+        trn2_model_modifications.append(
+            ModelConfigModifier.default_config().set(
+                target_config="model.decoder.transformer.layer.self_attention.attention."
```
A downside of representing these deeply nested configs as string paths is that they are brittle, and can quickly become outdated.

Have we considered using `cfg.visit` to achieve some of these modifications (e.g., https://github.com/apple/axlearn/blob/1c22688c1dc480777c43a1b8c8dd866dccda70e0/axlearn/common/layers.py#L266)?

(A bit late to review, so apologies if this discussion has already taken place.)

ruomingp commented on axlearn/common/config.py:
```diff
@@ -394,6 +394,66 @@ def set(self, **kwargs):
+        value = getattr(self, target_key)
+        return value.recursive_traverse(key_path[1:])
+
+    def get_recursively(self, key_path: Sequence[str]) -> Any:
```
```suggestion
    def get_recursively(self, path: Sequence[str]) -> Any:
```

ruomingp commented on axlearn/common/config.py:
```diff
@@ -394,6 +394,66 @@ def set(self, **kwargs):
+        """Recursively find the target key in the config and return its value.
+
+        Args:
+            key_path: A sequence of keys for indexing to get the target value.
```
Can path be empty? Maybe it can return `self` if path is empty?

apoorvtintin replied:
Added a check to see if path is empty, throws value error now if path is empty

ruomingp commented on axlearn/common/config.py:
```diff
@@ -394,6 +394,66 @@ def set(self, **kwargs):
+        """Recursively find the target key in the config and set its value.
+
+        Args:
+            key_path: A sequence of keys for indexing to set the target value.
```
Can path be empty?

ruomingp commented on axlearn/common/config.py:
```diff
@@ -394,6 +394,66 @@ def set(self, **kwargs):
+        Raises:
+            ValueError: A key in key_path is not found.
+        """
+        traverse_result = self.recursive_traverse(key_path)
```
Can we do something like:
```suggestion
        if not path:
            raise ValueError(...)
        parent = self.get_recursively(path[:-1])
```

apoorvtintin replied:
I removed the `recursive_traverse` method and added a value error if path is empty. I am still using path[1:] instead path[:-1] since we are traversing parents first.

apoorvtintin commented on axlearn/common/config.py:
```diff
@@ -394,6 +394,66 @@ def set(self, **kwargs):
+        # The key string.
+        key: str
+
+    def recursive_traverse(self, key_path: Sequence[str]) -> tuple[Any, str]:
```
Calling value.recursive_traverse() results in a pylint error "Access protected method from non member function". Let me see if I can avoid that somehow.

ruomingp replied:
I don't think we need this method. Please see my other comments.

apoorvtintin replied:
I removed this method

apoorvtintin commented on axlearn/common/trainer_config_modifier.py:
```diff
@@ -146,6 +137,110 @@ def __call__(self, cfg: SpmdTrainer.Config) -> SpmdTrainer.Config:
         return cfg
 
 
+class ModelConfigModifier(ConfigModifier):
```
The class makes some assumptions around the `klass` attribute of configs which are only present in modules, which is also what makes it non generic.

apoorvtintin replied:
@markblee I would love to get some more guidance on a name that fits this class better, do you have any suggestions?

markblee replied:
Hm, but all configs deriving from `Configurable` have `klass`. Did you mean `ModuleConfigModifier` perhaps, as it replaces arbitrary modules by path? (model -> module?)

apoorvtintin replied:
Yeah I agree `ModuleConfigModifier` is better, thanks!

apoorvtintin commented on axlearn/common/config.py:
```diff
@@ -394,6 +394,66 @@ def set(self, **kwargs):
+        key: str
+
+    def recursive_traverse(self, key_path: Sequence[str]) -> tuple[Any, str]:
+        """Recursively traverse the config to find the target key.
```
Both loops and recursion are fair choices I was following the pattern of `utils.get_recusively()` based on previous guidance.

apoorvtintin commented:
Thank you for the reviews @markblee @ruomingp @kelvin-zou!
@ruomingp I addressed most of your comments, please let me know if more changes are needed. @markblee We can discuss the usage of `cfg.visit` instead of using full module paths further, but I would like to address this in a follow up PR if possible.

ruomingp commented on axlearn/common/config.py:
```diff
@@ -394,6 +394,54 @@ def set(self, **kwargs):
+        Returns:
+            value at the path.
+        """
+        if not path:
+            raise ValueError("Path is empty.")
+
```
```suggestion
        if not path:
            return self

```

ruomingp commented on axlearn/common/config.py:
```diff
@@ -394,6 +394,54 @@ def set(self, **kwargs):
+        if len(path) == 1:
+            return child
+
+        return child.get_recursively(path[1:])
```
Nit: Can we avoid recursion with a loop?

ruomingp commented on axlearn/common/config.py:
```diff
@@ -394,6 +394,54 @@ def set(self, **kwargs):
+        if path[0] not in self:
+            raise ValueError(f"{path[0]} is not found in {self}.")
+
+        child = getattr(self, path[0])
+
+        if len(path) == 1:
+            return setattr(self, path[0], value)
```
```suggestion
        if len(path) == 1:
            return setattr(self, path[0], value)
        
        child = getattr(self, path[0])
``````

ruomingp commented on axlearn/common/config.py:
```diff
@@ -394,6 +394,54 @@ def set(self, **kwargs):
+        if len(path) == 1:
+            return setattr(self, path[0], value)
+
+        return child.set_recursively(path[1:], value=value)
```
Avoid recursion?

ruomingp commented on axlearn/common/config.py:
```diff
@@ -394,6 +394,54 @@ def set(self, **kwargs):
+        if not path:
+            raise ValueError("Path is empty.")
+
+        if path[0] not in self:
+            raise ValueError(f"{path[0]} is not found in {self}.")
```
Do we need this check? getattr and setattr will raise anyways.

markblee commented on axlearn/experiments/text/gpt/fuji.py:
```diff
@@ -434,7 +552,7 @@ def get_trainer_kwargs(
             learner_kwargs=dict(peak_lr=1.5e-4, weight_decay=0.1),
             max_sequence_length=max_sequence_length,
-            train_batch_size=train_batch_size,
+            train_batch_size=8,
```
Is this change intended?

markblee reviewed (commented):
Hi @apoorvtintin, please feel free leave `TODO(markblee):` for my comments and go ahead once @ruomingp approves. I'll be happy to send a follow-up PR to avoid blocking this one.

apoorvtintin commented:
Addressed all comments. @markblee I added TODOs for you to revisit cfg.visit later, please let me know if I should add more.

ruomingp commented on axlearn/common/config.py:
```diff
@@ -394,6 +394,58 @@ def set(self, **kwargs):
+            # TODO(markblee): maybe use cfg.visit instead of getattr
+            current = getattr(current, key)
+            index += 1
+
+            if index == len(path):
+                return current
```
```suggestion
```

apoorvtintin replied:
Thanks for the review @ruomingp. Can I get some more details here?

apoorvtintin replied:
Changed to `break`, let me know if you had something else in mind

ruomingp commented on axlearn/common/config.py:
```diff
@@ -394,6 +394,58 @@ def set(self, **kwargs):
+        if not path:
+            raise ValueError("Path is empty.")
+
+        current = self
+        for i, key in enumerate(path):
+            if i == len(path) - 1:
+                setattr(current, key, value)
+                return
+            else:
+                # TODO(markblee): maybe use cfg.visit instead of getattr
+                current = getattr(current, key)
```
```suggestion
        parent = self.get_recursively(path[:-1])
        setattr(current, path[-1], value)
```

apoorvtintin replied:
I moved to loop based on the previous comments, just to clarify we don't want loops for set_recursively but only for get_recursively?

apoorvtintin replied:
Got this, will make this change

apoorvtintin commented:
Thanks for the latest comments @ruomingp. I need some clarification on your latest comments. This PR is almost done, let's push this to the finish line! :)

kelvin-zou commented on axlearn/experiments/text/gpt/fuji.py:
```diff
@@ -151,6 +155,72 @@ def get_trainer_kwargs(
 
     rope_theta = ROPE_THETA[version]
 
+    # TRN2 specific model config modifications
```
Can we move all the modifications in a helper function? 
saying

def _generate_trainium2_custom_configs():
    ...
    return trn2_model_modifications, trn2_partition_spec_modifications

apoorvtintin replied:
Addressed it, lmk if it looks good

hanzhi713 commented on axlearn/experiments/text/gpt/fuji.py:
```diff
@@ -151,6 +155,72 @@ def get_trainer_kwargs(
+                },
+                # Sequence parallel shardings for norms.
+                "model.decoder.transformer.layer.self_attention.norm": {
+                    "input_partition_spec": ("fsdp", "model", None),
```
Why is there no "data"? Is it because trn2 don't support data and fsdp at the same time?

apoorvtintin replied:
Yes "data" over "fsdp" is supported. I'll make this change

kelvin-zou reviewed (commented):
Approve overall, please address @hanzhi713 's comment.

apoorvtintin commented:
Updated the PR to address failing tests, can we re-trigger the CI please? Thank you

apoorvtintin commented:
@kelvin-zou @hanzhi713 Thank you both for the review, I addressed all the comments. Let's merge this if it looks good

kelvin-zou reviewed (approved):
Thank you!

apoorvtintin commented:
Hello @ruomingp, can I please get an approval if this PR looks good? Looks like that is needed for the PR to merge.

kelvin-zou commented:
maybe @markblee  can have a second eye for the final round?

markblee commented on axlearn/common/config.py:
```diff
@@ -394,6 +394,49 @@ def set(self, **kwargs):
+            value at the path or self if path is empty.
+        """
+        current = self
+        index = 0
+
+        while path and index < len(path):
+            key = path[index]
+
+            # TODO(markblee): maybe use cfg.visit instead of getattr
+            current = getattr(current, key)
+            index += 1
+
+            if index == len(path):
+                break
```
Hm, did I miss something or can we just do:
```suggestion
        for part in path:
            current = getattr(current, part)
```

(In any case, I can make some changes in a follow-up.)

apoorvtintin replied:
Made this change, thanks @markblee

markblee reviewed (approved):
Approve to unblock.

markblee commented:
(Looks like we still need @ruomingp 's approval to unblock the 'requested changes'.)

apoorvtintin commented:
Hello @ruomingp @markblee looks like this PR green for merge, let me know if any more changes are needed.

apoorvtintin commented:
Rebased the PR to include latest Axlearn changes.

kelvin-zou commented:
Verified internally on upgrade PR, will merge it.
